PEM read-side refactor

kaduk opened the pull request:
WIP partially since we still have to make up our mind how strict we want the parser to be (and with flags, we even can offer multiple options).

See also RT4698

tlhackque commented:
My comments on RT4698 still apply - I don't quite understand the state change to "rejected but issue tracked on (this) PR".

Whitespace (anywhere) and odd line lengths are common UI issues.  Non-base64 chars have pluses and minuses.  For the former, consider an e-mail that gets quoted with '>' when forwarded.  The latter are described in the RFC.

As I mentioned in the RT, the OpenXPKI project has encountered this issue; see openxpki/openxpki/issues/437.

I maintain the Perl CPAN module Crypt::PKCS10, where I parse and output CSRs and public keys.

When parsing, I remove all whitespace, optionally remove non-base64 characters, and treat the result as one long string.   That's a lot simpler than insisting on 64 character lines (except the last).    And it accepts all valid input.  Of course, that's Perl.  But in C code it's easy to do the same thing by processing one character at a time.

Errors are best caught when parsing the asn.1 (and validating signatures), not by enforcing a strict punch-card era text format.  I recommend this.

For output, of course, I produce strictly compliant PEM.

In other applications, it's a bit more involved to detect the encryption header - but once past that, I use the same approach.

Bottom line, I encourage the Unix "accept liberal, produce conservative".  I'd settle for an option, but I don't think it's necessary - and OpenSSL has enough of those as is.

richsalz commented:
First off, I never said there weren’t interop concerns.  I am sorry if my quibbles/jokes over MUST sent the conversation down the wrong chute.

The ticket is closed because we’re moving to GitHub.  Please raise your concerns here, or perhaps Ben has addressed them.

tlhackque commented:
I understood the spirit of your comments, no problem.  I literally didn't understand why 'rejected' was the chosen state for the RT - but that's why I commented here.  'rejected' is usually a strong dismissive statement from a maintainer.

The original purpose of this PR was addressing refactoring and choice of memory pools, not the issue I raised.  I'm certainly OK with merging them for convenience; I'm grateful that Ben volunteered.

At first glance, Ben has re-worked the existing code, but not liberalized the parsing.  His comment at the top of this thread says "WIP", "we still need to decide..."

The code seems to have a lot of legacy and complexity; I haven't done a detailed review but applaud the courage in touching it.  If you take my advice, it could well get simpler.

I think my concerns are adequately listed here (incorporating by reference the RT thread).

I'll watch this thread and am willing to participate if that's useful.

richsalz commented:
I suppose I will use "resolved" in closing the others.  Thanks!

kaduk commented:
It is WIP largely because we need to decide what we want the behavior to be.
Having flags gives us options; there could be a flag for bug-for-bug compatibility, for ignoring all whitespace, all non-b64 characters, for enforcing line lengths, etc.
I would like some feedback from the team about which features are desired and what the defaults should be.
Unfortunately, I injured my hand yesterday and cannot type out the full list of open questions as I see it.

tlhackque commented:
Yes, flags provide options - but also complexity for users, maintenance and testing.

I settled on one option - whether to ignore non-b64.  While I don't think using them as a covert channel is a serious risk in OpenSSL data, it's a simple accommodation and sometimes catches corruption.  e.g. '=' in the middle of  a string can happen when someone copies quoted printable...

Whitespace is harmless.  So are random line lengths.

I don't have a case for bug-for-bug compatibility here.  Currently, valid data aborts processing.

I think I saw that OpenSSL will sometimes accept PEM data without the END label.  If someone relies on that, I'd have little sympathy.  That's the only case that comes to mind of OpenSSL's PEM processing being too liberal.

One could easily implement a separate tool for 'is this strict PEM?'.  Anyone who might want to validate that they're producing strict PEM would be better served by such a tool than by OpenSSL's idiosyncrasies...  And I think that reducing the options in OpenSSL code would be a good thing.

Of course once you accept liberal input, `openssl x -in foo -out bar` will produce strictly compliant PEM.

Hope your injury heals soon.

richsalz commented:
There is code in openssl that depends on some of the current behavior.  I forget details.  See the issue/PR from AGL that code a beer-mug in it.

FdaSilvaYY commented:
#1209

tlhackque commented:
Cute, but I know of no RFC that considers '-' as a comment character.  That's the 'covert channel' argument from the RFC.  Note that Google reports 'no ill effects' from its removal.

Most of the image could be retained with "ignore Base64" - only 7 distinct characters in the image are valid Base64.  oOCQCG0 There are plenty of other graphics with which to replace them.

Why would asn1parse care about the PEM?  It should get binary the same way that all PEM consumers do, and parse the binary.  If it doesn't, it should - if for no other reason than that as a debug parser, it should see the same bits as everything else.

richsalz commented:
Should doesn't always equal the way things work, and considerations of the deployed based.

tlhackque commented:
By 'should', I meant that if necessary, simplifying asn1parse to use the new input routines becomes part of this work - or maybe it's a separate bug against asn1parse.

A diagnostic tool can be evil if it sees different bits from the real consumer.

Is the '-' comment line behavior documented?  Is that really what allows the image - it looks as though it may be relying on the 'if the line is not 64 characters long, the PEM is over' logic?

Anyhow, my 3 cents is that if someone relies on this, it's not a hardship to deal with it externally
e.g.  `sed -e'/^-/d' file.pem | openssl foo`

Compatibility is important - but it's always a trade-off.  

I've said enough.  I'll yield to your team's judgement - my primary concern is that OpenSSL accept RFC-compliant input.

richsalz commented:
what's going on with this now?

kaduk commented:
I timed out on waiting for feedback about what flags the dev team was interested in seeing, and was going to default to implementing what @tlhackque wanted and the secure memory flag and remove the WIP label, but then started blocking on time to actually do that work.
It seems somewhat unlikely that time to do that work will appear this week while I'm at the IETF.

richsalz commented:
Might be worth trying to start discussion on the tls-dev list and perhaps close this.

levitte commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,256 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
-        return (0);
+    if (flags & PEM_FLAG_WEAK_EOL) {
+        /* Strip trailing whitespace */
+        while ((len >= 0) && (linebuf[len] <= ' '))
```
What about tabs?  Maybe use an appropriate ctype function...  `isspace()` or `isblank()`?

kaduk replied:
This part is for "bug-for-bug" compatibility with the historical code, and is in fact part of what inspired me to do the wholesale refactor.

tlhackque replied:
Note also that  <= ' ' in ASCII includes all control characters: NUL (000) thru US (0x1F).  So it does include tab [HT].  (and FF, and VT and CR and LF ...)  In EBCDIC, it also happens to do more or less the same thing, though it's a toss-up as to whether that was in the mind of the original author...

ctype would be the right thing for new (sane) code that wants to detect whitespace.

kaduk commented:
Rebased to lateset master, with some updates: renamed flag from "weak eol" to "eay compatible", since that's what it's intended to be, and added a b64-only (i.e., "strict") option, with the default behavior being to treat control characters as whitespace (i.e., ignored) ... which means that ASCII non-base64 characters get passed through to EVP_Decode, for now.
This is all up for discussion, of course, but I removed the WIP label since I am actually proposing this change.

kaduk commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,288 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+/* Some helpers for PEM_read_bio_flags(). */
+
+/* TODO Presumably this should be a table lookup, but are these the right ranges? */
+static int isb64(char c)
```
This TODO should be resolved before this gets merged ... thoughts?

levitte replied:
You might want to have a look in `crypto/evp/encode.c`...

kaduk replied:
I wasn't super keen on exposing that stuff to the whole library and leaving it in encode.c, nor was I super enthusiastic about splitting it out into a more generic place.  But, if that's how it has to be, I can do it...

tlhackque replied:
Not sure why you would want a table; in all codesets those ranges are contiguous.  The issues come when you assume the relative placements of upper vs lower case, or digits vs letters, or where other graphic (e.g. + and /) appear relative to other characters.

Not that this is particularly performance sensitive, but a table (256 bytes? bigger?) would require memory and data cache; the range tests are compact (minimum instruction and no data cache impact) and cover the cases.  I don't think a table adds clarity to the code.  The ranges capture it.  This isn't crypto, so no reason to care about constant time.  So, I'd stick with what you have.

You might return -1 for invalid, and the encoded value for OK (e.g. 0 for 'A' & 26 for 'a'); that would let the caller do something like `if( (nibble = isb64(in)) < 0 ) return error (or ignore); val |= nibble << state;`  It's easy to do: `if( c >= 'a' && c <= 'z' ) return (c - 'a') + 26;` etc.

You might want a state machine somewhere.; that might work as a (small) table.  (I haven't reviewed the whole commit, and don't plan to.  But noticed the questions.)

The ranges look right to me, assuming you handle pad ('=') elsewhere.  Ref: [https://tools.ietf.org/html/rfc4648#page-5](rfc4648), which  has a table.

My inclination for default is to be lax; interpret B64 chars, ignore everything else.  Pad only at the end, and only the permissible cases (0,1,2).  There's a case for exception if non-B64 other than whitespace (which is sp,ht,lf,cr,ff, & often vt; sometimes NUL) or proper padding.  The EAY line length-sensitive weirdness is at best a desperate fallback for compatibility; should not be a default.  (Even though it is old behavior, it causes reasonable input to be rejected.)

I'm assuming that the apps / consumers will be updated as needed to resolve the external manifestations. Hopefully a consistent option: `-pemfmt [obsolete|strict]` with default = `lax` ?

Hope this helps.  Thanks for working on it.

kaduk commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,288 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+    *data = NULL;
+    if ((flags & PEM_FLAG_EAY_COMPATIBLE) && (flags & PEM_FLAG_ONLY_B64)) {
+        /* These two are mutually incompatible; bail out. */
+        PEMerr(PEM_F_PEM_READ_BIO_FLAGS, ERR_R_PASSED_INVALID_ARGUMENT);
```
Is this a reasonable error reason?

levitte replied:
I think so

kaduk commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,288 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+
+    EVP_DecodeInit(ctx);
+    /* Not sure which is more appropriate: BIO_get_mem_ptr() or BIO_get_mem_data() */
+    BIO_get_mem_ptr(dataB, &buf_mem);
```
Input requested on which BIO_get routine is better to use here.

levitte replied:
Both do more or less the same thing, the crucial thing is how you want the data.

levitte commented:
I've just had another look at this work (prompted by @kaduk's comment in #2746)...  and I wonder, are there thoughts of having the resulting EVP_PKEY et al from PEM_read_bio_PrivateKey and friends in secure memory as well?

kaduk commented:
> are there thoughts of having the resulting EVP_PKEY et al from PEM_read_bio_PrivateKey and friends in secure memory as well?

Probably makes sense as a later addition ... there are several codepaths that could be dispatched to by that function and it's not clear that it would need to be done at the same time.  Thanks for pointing it out!

levitte commented:
Later addition is fine, I was only asking about the intention going forward.

kaduk commented:
Rebased and fixed a couple needless behavior changes exposed by 04-test_pem.t.

levitte commented on crypto/pem/pem_lib.c:
```diff
@@ -228,28 +228,47 @@ static int check_pem(const char *nm, const char *name)
+    return OPENSSL_malloc(num);
+}
+
+static int PEM_bytes_read_bio_flags(unsigned char **pdata, long *plen,
```
Internal functions should have lower case prefix

levitte commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,286 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+        while ((len >= 0) && (linebuf[len] <= ' '))
+            len--;
+        linebuf[++len] = '\n';
+        linebuf[++len] = '\0';
```
This is repeated in all condition branches, so I's suggest replacing these two lines with `len++;`...

levitte commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,286 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+        /* The caller allocated LINESIZE+1, so this is safe. */
+        linebuf[i] = '\n';
+        linebuf[i + 1] = '\0';
+        len = i + 1;
```
These four lines with `len = i;`...

levitte commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,286 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+        /* The caller allocated LINESIZE+1, so this is safe. */
+        linebuf[i] = '\n';
+        linebuf[i + 1] = '\0';
+        len = i + 1;
```
These four lines with `len = i;`...

levitte commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,286 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+        linebuf[i] = '\n';
+        linebuf[i + 1] = '\0';
+        len = i + 1;
     }
```
And finally insert these two lines here:

``` C
    linebuf[len++] = '\n';
    linebuf[len] = '\0';
```

levitte commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,286 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
-            nohead = 1;
-            break;
+        /* Else, a line of text -- could be header or data; we don't
+         * know yet.  Just pass it through. */
```
Our C block comments should look like this:

``` C
/*
 * text
 * text
 */
```

levitte commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,286 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+    size_t len;
+
+    /* Need to hold trailing NUL (accounted for by BIO_gets() and the newline
+     * that will be added by sanitize_line() (the extra '1'). */
```
We'd like our block comments to look like this:

``` C
/*
 * text
 * text
 */
```

levitte commented on include/openssl/pem.h:
```diff
@@ -236,6 +236,16 @@ int PEM_do_header(EVP_CIPHER_INFO *cipher, unsigned char *data, long *len,
+int PEM_bytes_read_bio_secmem(unsigned char **pdata, long *plen, char **pnm,
+                              const char *name, BIO *bp, pem_password_cb *cb,
+                              void *u);
+#   endif
```
Move this one down with `PEM_bytes_read_bio`

levitte commented:
Apart from my remarks of the changes, you need to fix the documentation further, as noted by find-doc-nits:

```
(cd .; /usr/bin/perl util/find-doc-nits -n ) >doc-nits
if [ -s doc-nits ] ; then cat doc-nits; rm doc-nits ; exit 1; fi
doc/man3/PEM_bytes_read_bio.pod:1: missing copyright
doc/man3/PEM_read_bio_flags.pod:1: PEM_FLAG_SECURE missing from NAME section
doc/man3/PEM_read_bio_flags.pod:1: PEM_FLAG_EAY_COMPATIBLE missing from NAME section
doc/man3/PEM_read_bio_flags.pod:1: PEM_FLAG_ONLY_B64 missing from NAME section
doc/man3/PEM_read_bio_flags.pod:1: missing copyright
doc/man3/PEM_read_bio_flags.pod:1: period in NAME section
make: *** [doc-nits] Error 1
```

Also, you need to rebase on a fresher master

kaduk commented:
Looking at the commits again caused me to feel uneasy about the lack of test coverage for PEM_FLAG_ONLY_B64, which turns out to have been justified.
As it stands, that flag is not usable with encrypted files, though it would take some thinking to determine if ther's a way to fix that that does not involve a second pass over things.

levitte commented:
> Looking at the commits again caused me to feel uneasy about the lack of test coverage for PEM_FLAG_ONLY_B64, which turns out to have been justified.
> As it stands, that flag is not usable with encrypted files, though it would take some thinking to determine if ther's a way to fix that that does not involve a second pass over things.

Depends what you mean with "encrypted".  If "encrypted by PEM headers", then yes, you're right, but if it's encrypted in a PKCS#8 container, then PEM_FLAG_ONLY_B64 should be possible to use, no?

kaduk commented:
"encrypted with PEM headers", yes.

levitte commented:
There are several ways you can do that...  one idea is that you can sanitize with `flags & (found_header ? ~0 : ~PEM_FLAG_ONLY_B64)`, and you can also detect if your first body line is a header by looking for a colon (:), and if not, allow sanitize to see PEM_FLAG_ONLY_B64 already then.

levitte commented:
You seem to have upset the CIs.  Most importantly, the Appveyor ones show this:

```
Error processing -s_key keyU.ss
804:error:0909006C:PEM routines:get_name:no start line:..\crypto\pem\pem_lib.c:776:Expecting: ANY PRIVATE KEY
804:error:140B0009:SSL routines:SSL_CTX_use_PrivateKey_file:PEM lib:..\ssl\ssl_rsa.c:550:
ssltest_old.exe -s_key keyU.ss -s_cert certU.ss -c_key keyU.ss -c_cert certU.ss -bio_pair => 1
```

But only when building shared!

levitte commented on test/pemtest.c:
```diff
@@ -0,0 +1,77 @@
+{
+    BIO *b = BIO_new(BIO_s_mem());
+    char *name, *header;
+    unsigned char *data;
```
Make sure to initialise `name`, `header` and `data`.

kaduk commented:
It may be hard to track down the AppVeyor issue without a windows machine; the file it's complaining about is generated during the build with "..\apps\openssl.exe req -config ..\..\test\Uss.cnf -out reqU.ss -keyout keyU.ss -new > err.ss".
The error indicates that libcrypto is getting an error or zero-length input when trying to read the first line of the file to look for the "-----BEGIN FOO-----" name line.
But it looks like this code is doing essentially the same operation as the historical code -- BIO_gets() and checking if the return value is less than or equal to zero.
I guess I'll push my other changes and give it another crack, at least.

levitte commented:
I'll have a look on a VMS system, that can sometimes trigger the same kind of faults as Windows...

kaduk commented:
Thanks for trying it out on VMS.

FWIW, I put together a prototype for what it could look like to use ':' to detect the PEM header, at https://github.com/kaduk/openssl/commit/03434ffaec6c5cd7cddbfef147afc763c4c5df76 (my 'pem2' branch).  But I didn't dig back into RFC 1421 to confirm that it should always work, so it's not part of this pull request.

levitte commented:
> But I didn't dig back into RFC 1421 to confirm that it should always work

From RFC 1421:

```
       Pre-Encapsulation Boundary (Pre-EB)
           -----BEGIN PRIVACY-ENHANCED MESSAGE-----

       Encapsulated Header Portion
           (Contains encryption control fields inserted in plaintext.
           Examples include "DEK-Info:" and "Key-Info:".
           Note that, although these control fields have line-oriented
           representations similar to RFC 822 header fields, the set
           of fields valid in this context is disjoint from those used
           in RFC 822 processing.)

       Blank Line
           (Separates Encapsulated Header from subsequent
           Encapsulated Text Portion)

       Encapsulated Text Portion
           (Contains message data encoded as specified in Section 4.3.)

       Post-Encapsulation Boundary (Post-EB)
           -----END PRIVACY-ENHANCED MESSAGE-----


                   Encapsulated Message Format
                            Figure 1
```

Note "line-oriented representations similar to RFC 822", which indicates *key*:*value* form of lines.  The first line *must* have a colon to be such a header.

levitte commented:
Unfortunately, VMS wasn't much of a help...

@dot-asm, do you have the possibility to check what goes wrong with this?

(@kaduk, you will have to be patient...  Easter is a 4-day holyday in Sweden, Friday through Monday, so don't expect too much from me)

kaduk commented:
Managed to scrounge up a windows VM with which to reproduce.
The error message is somewhat confusing, as it is generated after scanning through the entire file looking for a start line and hitting EOF.

Somehow, after sanitize_line() we are ending up with spurious whitespace at the end of the line, so that we are comparing ```"---- \n"``` to TAILSTR (i.e., ```"-----\n"```), which unsurprisingly fails.  More investigation and (presumably) updated patch pending...

kaduk commented:
In retrospect, this is an unsurprising natural consequence of CRLF line endings, and conditionals that explicitly compare against '\n'.  The extra '\r' gets treated as a control character and forced to whitespace, which is then trailing whitespace at the end of the line and causes the check to fail.

kaduk commented:
Oh, appveyor needs a rebase before it will even try building?  Interesting.

levitte commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +694,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+        /* EVP_DecodeBlock strips leading and trailing whitespace, so just strip
+         * control characters in-place and let everything through. */
+        for (i = 0; i < len; ++i) {
+            if (linebuf[i] == '\n')
```
You might want to check for `\r` as well...

levitte replied:
Ah, seems like you figured that out...

kaduk commented:
It looks like I may need to push the squashing off until tomorrow, but we can let appveyor have a crack at the new code at least.

kaduk commented:
Hmm, the pyca tests are failing, in a way plausibly related to how this change makes the format requirements for private keys slightly more strict.  How much do we care?

richsalz commented:
We care very much. Breaking what used to work is definitely not allowed in this release.

kaduk commented:
Hmm, I think it's the same 3 PYCA tests that failed before and after adding PEM_FLAG_EAY_COMPATIBLE to the PrivateKey loading (via PEM_bytes_read_bio_secmem()), but didn't take notes on the original travis run.

Just to be clear, the failing cases look to be named "test_corrupt_traditional_format_der", "test_wrong_private_format", and "test_wrong_public_format", which is why I'm not 100% convinced that this is a case where we need to retain bug-for-bug compatibility.  But, I don't think I'll be able to say more about why these cases are failing without running them locally, which will take some time (unless @jon-oracle decides to jump in, but I do not plan to wait for that to happen).

kaduk commented:
Ah, just having the pyca source to poke at is helpful, in that it lets me understand what the travis output is trying to tell me.  The test runs
```
        with raises_unsupported_algorithm(
                _Reasons.UNSUPPORTED_PUBLIC_KEY_ALGORITHM):
>           load_pem_public_key(key_data, backend)
``` (the ```>``` indicates that's where we are in the call chain, and the other stuff is context), indicating that it expects to get a "Unsupported public key algorithm" exception, as this operation is in fact supposed to fail.  But, the code to handle errors from the OpenSSL call is doing:
```
        elif errors[0][1:] in (
            (
                self._lib.ERR_LIB_ASN1,
                self._lib.ASN1_F_ASN1_CHECK_TLEN,
                self._lib.ASN1_R_WRONG_TAG
            ),
            (
                self._lib.ERR_LIB_PEM,
                self._lib.PEM_F_PEM_READ_BIO,
                self._lib.PEM_R_NO_START_LINE
            ),
        ):
            raise UnsupportedAlgorithm(
                "Unsupported public key algorithm.",
                _Reasons.UNSUPPORTED_PUBLIC_KEY_ALGORITHM
            )
```
that is, inspecting the error library, function and reason code.  But, the function code has changed, since the function is now PEM_read_bio_flags(), so this check does not trigger, and things fall through to the else clause that just does ```raise ValueError("Could not deserialize key data.")```, which is not the expected exception in the test function, so is counted as a failure.

So, I guess it's time to address the elephant in the room: are error function codes part of the stability guarantee?  Attempting to make them be part of it would seem to run into direct conflict with ```make update```, which insists that the function code matches the name of the function it's being used in; that in turn would prevent us from making an existing function into a wrapper around a new function with extended functionatliy.  But probably this discussion is better held on openssl-dev than in an already-long github comment thread; I'll work on drafting an email.

levitte commented:
> So, I guess it's time to address the elephant in the room: are error function codes part of the stability guarantee? Attempting to make them be part of it would seem to run into direct conflict with make update, which insists that the function code matches the name of the function it's being used in; that in turn would prevent us from making an existing function into a wrapper around a new function with extended functionatliy. But probably this discussion is better held on openssl-dev than in an already-long github comment thread; I'll work on drafting an email.

Well, I think that most of the time, we've considered the library and reason code combo to be reasonably stable, but the function code can vary.  I'd say that the pyca stuff is *overly* sensitive in its checks of error data.

levitte commented:
... on the other hand, who are we to say how others use our public stuff?

levitte commented:
But yeah, all in all, I'd like to consider the check of the function code a bit over the top...

levitte commented:
I just had a look at the [pyca/cryptography](https://github.com/pyca/cryptography), and it looks like they have updated their code, the checks in `_handle_key_loading_error` are different.  Maybe we should update our sub-module...

levitte commented:
So how about this, either patch your branch with the contents of #3346 and see if that helps you, or wait 'til #3346 has been merged, rebase on top of that and try again?

levitte commented:
#3346 is now merged, feel free to rebase on a fresher master and see if that helps matters...

kaduk commented:
> #3346 is now merged, feel free to rebase on a fresher master and see if that helps matters...

Having compared the pyca sources before and after, it looks like it should help.
Running a local test build now before pushing...

kaduk commented:
Sigh, pyca/cryptography/vectors/cryptography_vectors/asymmetric/Traditional_OpenSSL_Serialization/key1.pem has trailing whitespace on the "END RSA PUBLIC KEY" line, so I cannot sneak in the lack of PEM_FLAG_EAY_COMPATIBLE for PEM_read_bio_secmem(), alas.

levitte commented:
Yeah, enhanced strictness will have to wait... Also, not sure we should ever. "Be generous in what you accept" 'n all that.

richsalz commented:
Requiring a new flag to be used to parse old files that used to work is not acceptable.
A PEM_PARSE_STRICT flag would be okay, for example.

kaduk commented:
@richsalz with the latest commit, the behavior of all previously existing public APIs remains unchanged.  If you use the  *new* public API and do not specify PEM_FLAG_EAY_COMPATIBLE, you get the more-strict behavior.  Does that suffice for your acceptibility criteria?
(I would like to remove the EAY-compatible stuff in some future major release, but clearly that is not okay for 1.1.1.  Most likely path would be to drop the flag from the aforementioned previously existing public API's calls to the _flags variant for 1.2, so people needing the behavior could switch to calling the flags variant directly as a transition plan, then remove the compat behavior entirely in 1.3)

richsalz commented:
As for the elephant, remember when I tried to remove unused function codes and we decided we can't?

kaduk commented:
Removing the function codes entirely would break people's build.
Just not returning them anymore is somewhat different.

richsalz commented on doc/man3/PEM_bytes_read_bio.pod:
```diff
@@ -0,0 +1,80 @@
+
+=head1 DESCRIPTION
+
+PEM_bytes_read_bio() reads PEM-formatted data from the BIO I<bp> for the
```
RFC reference?

richsalz commented on doc/man3/PEM_read_bio_flags.pod:
```diff
@@ -0,0 +1,70 @@
+ #define PEM_FLAG_SECURE             0x1
+ #define PEM_FLAG_EAY_COMPATIBLE     0x2
+ #define PEM_FLAG_ONLY_B64           0x4
+ int PEM_read_bio_flags(BIO *in, char **name, char **header,
```
we tend to use ```_ex``` suffix for this kind of new API.

kaduk commented:
(The many red 'x'es were due to a change in my local copy that didn't make it into a commit.)
Rebased, squashed, and attempted to take all feedback so far.

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -228,28 +228,43 @@ static int check_pem(const char *nm, const char *name)
-                       void *u)
+static void pem_flag_free(void *p, unsigned int flags)
+{
+    if (flags & PEM_FLAG_SECURE) {
```
remove the curly and the return and just use if/else.

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -228,28 +228,43 @@ static int check_pem(const char *nm, const char *name)
+
+static void *pem_flag_malloc(int num, unsigned int flags)
+{
+    if (flags & PEM_FLAG_SECURE)
```
use the ?: operator

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -228,28 +228,43 @@ static int check_pem(const char *nm, const char *name)
-    for (;;) {
-        if (!PEM_read_bio(bp, &nm, &header, &data, &len)) {
+    do {
+        pem_flag_free(nm, flags);
```
these are new functions?  get rid of the ```_flag``` part from the name.

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
-    BUF_MEM *nameB;
-    BUF_MEM *headerB;
-    BUF_MEM *dataB, *tmpB;
+    if (c >= 'A' && c <= 'Z')
```
use the ctype functions and then this probably becomes a define:
```
/* NOTE |c| is evaluated multiple times! */
#define b64(c) (islanum(c) || c == '+' || c == '/' || c == '=')
```

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
-        goto err;
+#define LINESIZE 255
+/* Note trailing spaces for begin and end. */
+static char const beginstr[] = "-----BEGIN ";
```
are you sure that shouldn't be "const char" ?

kaduk replied:
There is no semantic difference between "const char" and "char const", only stylistic.

richsalz replied:
good.  then use 'const char' which we use everywhere but once:
```
; g grep 'char const' | grep -v constant_
crypto/evp/bio_md.c:static int md_write(BIO *h, char const *buf, int num);
```

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+        len = sanitize_line(linebuf, len, flags & ~PEM_FLAG_ONLY_B64);
+
+        /* Allow leading empty or non-matching lines. */
+    } while (strncmp(linebuf, beginstr, BEGINLEN) != 0 || len < TAILLEN ||
```
I think you should put a linebreak before each ```||``` to show they are separate tests.  Okay if you don't.

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+        return 0;
+    }
+
+    for (line = 0;;line++) {
```
spave after each simicolon.

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+    for (line = 0;;line++) {
+        flags_mask = ~0u;
+        len = BIO_gets(bp, linebuf, LINESIZE);
+
```
if the len-length test up to the len assignment, and move that blank line before 835's if.

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
-            memcpy(nameB->data, &(buf[11]), i - 6);
-            nameB->data[i - 6] = '\0';
-            break;
+            got_header = 2;
```
rather than magic numbers, use define's or, probably better, an enum for got_header.

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+    /* Not sure which is more appropriate: BIO_get_mem_ptr() or BIO_get_mem_data() */
+    BIO_get_mem_ptr(dataB, &buf_mem);
+    len = buf_mem->length;
+    if (EVP_DecodeUpdate(ctx, (unsigned char*)buf_mem->data, &len, (unsigned char*)buf_mem->data, len) < 0 ||
```
move the ```||``` to the next line with the evp-decode call.  and indent that call an extra shift-width (4)

kaduk replied:
Done

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
-    if (i < 0) {
-        PEMerr(PEM_F_PEM_READ_BIO, PEM_R_BAD_BASE64_DECODE);
-        goto err;
+    if (flags & PEM_FLAG_SECURE)
```
use the ?: operator.

kaduk replied:
Done.

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+    dataB = BIO_new(bmeth);
+    if (headerB == NULL || dataB == NULL) {
+        PEMerr(PEM_F_PEM_READ_BIO_EX, ERR_R_MALLOC_FAILURE);
+        goto cleanup;
```
we almost never use 'cleanup' as a goto target.  it's either end or err more than 90% of the time.

kaduk replied:
Okay.  Went with 'end', as the normal/success control flow also goes through the code after the label, which makes it not exactly the error-handling as would be implied by 'err'.

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+    *len_out = 0;
+    *name_out = *header = NULL;
+    *data = NULL;
+    if ((flags & PEM_FLAG_EAY_COMPATIBLE) && (flags & PEM_FLAG_ONLY_B64)) {
```
memleak of ctx.  use a goto.  or move this test before the ctx assignment.

richsalz commented on crypto/pem/pem_lib.c:
```diff
@@ -661,177 +690,299 @@ int PEM_read(FILE *fp, char **name, char **header, unsigned char **data,
+    len += taillen;
+    buf_mem->length = len;
+
+    if (len == 0)
```
how can this be true?  taillen is negative?

kaduk replied:
If both the starting len and taillen are zero, i.e., there was no content in the PEM file, just the BEGIN/END lines.
Mostly, IIRC, it was intended to avoid malloc(0), but I originally wrote this quite some time ago and could be misremembering.

richsalz replied:
add a comment for that.  assuming you do so, and since someone else will review, +1 from me.

kaduk commented:
And rebased again so the CI will actually take a look.

levitte commented:
This should be taken out of WIP at this point, no?

kaduk commented:
> This should be taken out of WIP at this point, no?

I could have sworn that I already had done so; maybe it reverted back somehow (on rebase?).
Regardless, removed WIP from the title.

kaduk commented:
Ah, the history feed has

     richsalz changed the title from PEM read-side refactor to WIP PEM read-side refactor 6 days ago

Shrug.

kaduk commented:
@levitte it seems that Rich is happy now, and the WIP is gone; can you take another look?

levitte commented:
@richsalz, wanna merge or shall I?

richsalz commented:
go  ahead.

levitte commented:
Merged.  204afd81b12c71d625e89599c0eef33588afc1f0 to fa3ed5b2c2b508a6444124fdf12ecbb4898007ed
